File: java-core-plugins/src/main/java/org/unlaxer/tramli/plugins/idempotency/IdempotencyRuntimePlugin.java
package org.unlaxer.tramli.plugins.idempotency;

import org.unlaxer.tramli.FlowEngine;
import org.unlaxer.tramli.plugins.api.PluginDescriptor;
import org.unlaxer.tramli.plugins.api.RuntimeAdapterPlugin;

public final class IdempotencyRuntimePlugin implements RuntimeAdapterPlugin<IdempotentRichResumeExecutor> {
    private final IdempotencyRegistry registry;

    public IdempotencyRuntimePlugin(IdempotencyRegistry registry) {
        this.registry = registry;
    }

    @Override
    public PluginDescriptor descriptor() {
        return new PluginDescriptor("idempotency", "Idempotency", "Binds a FlowEngine to duplicate-suppression helpers.");
    }

    @Override
    public IdempotentRichResumeExecutor bind(FlowEngine engine) {
        return new IdempotentRichResumeExecutor(engine, registry);
    }
}

File: ts-plugins/src/idempotency/idempotency-runtime-plugin.ts
import type { FlowEngine } from '@unlaxer/tramli';
import type { RuntimeAdapterPlugin, PluginDescriptor } from '../api/types.js';
import type { IdempotencyRegistry } from './types.js';
import { IdempotentRichResumeExecutor } from './idempotent-rich-resume-executor.js';

export class IdempotencyRuntimePlugin implements RuntimeAdapterPlugin<IdempotentRichResumeExecutor> {
  constructor(private readonly registry: IdempotencyRegistry) {}

  descriptor(): PluginDescriptor {
    return {
      id: 'idempotency',
      displayName: 'Idempotency',
      description: 'Binds a FlowEngine to duplicate-suppression helpers.',
    };
  }
  kind() { return 'RUNTIME_ADAPTER' as const; }

  bind(engine: FlowEngine): IdempotentRichResumeExecutor {
    return new IdempotentRichResumeExecutor(engine, this.registry);
  }
}
